File: templates/private-repo/scripts/compose.py
#!/usr/bin/env python3
"""
组合脚本：从 data/ 信息全集按 scripts/variants.yml 配方生成 resumes/*.yml

用法：
    python scripts/compose.py            # 生成所有变体
    python scripts/compose.py frontend   # 只生成指定变体
    python scripts/compose.py --check    # 只做完整性检查，不写文件

原则：
- data/ 是唯一真相源，本脚本只读 data/，生成的 resumes/*.yml 不应手改
- 组稿键（id / tags / 以 _ 开头）不进入最终简历；notes 仅作旧数据兼容并剥除
- achievements 每条成就可打 tags；无标签成就视为通用，任何方向都保留
"""
import json
import os
import re
import sys

import yaml

from font_options import get_typography_font_family, normalize_font_settings
from brand_icons import detect_url_brand

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.path.join(ROOT, "data")
VARIANTS_FILE = os.path.join(ROOT, "scripts", "variants.yml")
OUT_DIR = os.path.join(ROOT, "resumes")

META_KEYS = {"id", "tags", "notes"}
ACHIEVEMENTS_KEY = "achievements"

HEADER = (
    "# 本文件由 scripts/compose.py 自动生成，请勿手改。\n"
    "# 改信息 -> data/；改筛选/模板/章节顺序 -> scripts/variants.yml\n"
    "---\n"
)


class Dumper(yaml.SafeDumper):
    """null 输出为空值（endDate: 表示"至今"）"""


def _null_representer(dumper, _):
    return dumper.represent_scalar("tag:yaml.org,2002:null", "")


Dumper.add_representer(type(None), _null_representer)


def load(path):
    with open(path, encoding="utf-8") as f:
        return yaml.safe_load(f) or {}


def strip_meta(item):
    """去掉元数据键，返回纯简历字段"""
    if not isinstance(item, dict):
        return item
    # subtitle/source/role/stage 为管理端字段，schema 无此字段，不进简历
    return {k: v for k, v in item.items() if k not in META_KEYS and k not in ("subtitle", "source", "role", "stage") and not k.startswith("_")}


def tag_overlap(item_tags, wanted_tags):
    if not wanted_tags:
        return True
    if not item_tags:
        return False
    return bool(set(item_tags) & set(wanted_tags))


def select_entries(entries, cfg):
    """按配方选择条目：include-all / ids / tags"""
    if not entries:
        return []
    if cfg is None or cfg.get("include") == "all":
        return list(entries)
    ids = cfg.get("ids")
    tags = cfg.get("tags")
    result = []
    for e in entries:
        if ids is not None:
            if e.get("id") in ids:
                result.append(e)
            continue
        if tag_overlap(e.get("tags"), tags):
            result.append(e)
    return result


def collect_achievements(entry, wanted_tags):
    """成就点 -> markdown 列表（按标签过滤）；无成就点返回 None"""
    if ACHIEVEMENTS_KEY not in entry:
        return None
    items = []
    for a in entry[ACHIEVEMENTS_KEY]:
        if isinstance(a, dict):
            # 无标签成就 = 通用，保留；有标签的必须与方向命中
            if wanted_tags and a.get("tags") and not tag_overlap(a.get("tags"), wanted_tags):
                continue
            if a.get("text"):
                items.append(str(a["text"]))
        else:
            items.append(str(a))
    if not items:
        return None
    return "\n".join("- " + t for t in items)


def build_summary(entry, wanted_tags):
    """成就点 -> summary（按标签过滤）；无成就点则保留原 summary"""
    if ACHIEVEMENTS_KEY not in entry:
        return entry.get("summary")
    return collect_achievements(entry, wanted_tags)


def to_summary_string(summary):
    """summary 规范化为 markdown 字符串（schema 要求 summary 是字符串）"""
    if summary is None:
        return None
    if isinstance(summary, str):
        return summary
    if isinstance(summary, list):
        lines = ["- " + str(s) for s in summary if s]
        return "\n".join(lines) if lines else None
    return str(summary)


def normalize_project_summary(summary):
    text = (to_summary_string(summary) or "").strip()
    if not text:
        return None
    # 已使用 Markdown 列表时原样保留缩进；旧纯文本按行迁移为顶级项目要点。
    if re.search(r"^\s{0,3}(?:[-+*]|\d+[.)])\s+", text, re.MULTILINE):
        return text
    lines = [line.strip() for line in text.splitlines() if line.strip()]
    return "\n".join("- " + line for line in lines) if lines else None


def normalize_project_background(description):
    text = str(description or "").strip()
    if not text:
        return None
    return text if re.match(r"^项目背景[：:]", text) else "项目背景：" + text


def compact_skills(items):
    # 按细分方向（tags）分组：每个方向一行“方向：技能、技能”，跨方向技能同时出现在多行。
    dir_order = []
    groups = {}
    for item in items:
        candidates = item.get("keywords") if isinstance(item.get("keywords"), list) and item.get("keywords") else [item.get("name")]
        names = []
        for candidate in candidates:
            name = str(candidate or "").strip()
            if name.startswith("熟悉 ") or name.startswith("掌握 "):
                name = name[3:].strip()
            if name and name not in names:
                names.append(name)
        if not names:
            continue
        tags = [t for t in (item.get("tags") or []) if isinstance(t, str) and t.strip()]
        dirs = tags if tags else ["通用"]
        for d in dirs:
            if d not in groups:
                groups[d] = []
                dir_order.append(d)
            for n in names:
                if n not in groups[d]:
                    groups[d].append(n)
    return [
        {"name": f"{d}：{'、'.join(groups[d])}", "level": "Expert", "keywords": []}
        for d in dir_order
        if groups[d]
    ]


def compact_interests(items):
    names = []
    for item in items:
        name = str(item.get("name") or "").strip()
        if name and name not in names:
            names.append(name)
    return [{"name": "、".join(names), "keywords": []}] if names else []


def compose_basics():
    b = strip_meta(load(os.path.join(DATA_DIR, "basics.yml")))
    # photo 是管理端构建元数据；YAMLResume basics schema 不接受该字段，由后处理注入。
    b.pop("photo", None)
    if isinstance(b.get("summary"), list):
        b["summary"] = to_summary_string(b["summary"])
    return b if b else None


def compose_list(block, cfg):
    entries = load(os.path.join(DATA_DIR, block + ".yml")) or []
    selected = select_entries(entries, cfg)
    if not selected:
        return None
    wanted = (cfg or {}).get("tags")
    out = []
    for e in selected:
        item = strip_meta(e)
        # 字段别名：work 章节的 schema 字段名是 name（数据里用 company 更自然）
        if "company" in item and "name" not in item:
            item["name"] = item.pop("company")
        if block == "projects":
            # 字段映射：数据 background → schema description；数据 tech → schema keywords
            if "background" in item:
                item["description"] = item.pop("background")
            if "tech" in item:
                item["keywords"] = item.pop("tech")
            summary = normalize_project_summary(e.get("summary"))
            achievements = collect_achievements(e, wanted)
            parts = []
            if summary:
                parts.append(summary)
            if achievements:
                parts.append("**成果**\n" + achievements)
            if parts:
                item["summary"] = "\n\n".join(parts)
            else:
                item.pop("summary", None)
            item.pop(ACHIEVEMENTS_KEY, None)
            background = normalize_project_background(item.get("description"))
            if background:
                item["description"] = background
            else:
                item.pop("description", None)
        else:
            summary = build_summary(e, wanted)
            if summary is not None:
                item["summary"] = to_summary_string(summary)
            elif ACHIEVEMENTS_KEY in e:
                item.pop("summary", None)
            item.pop(ACHIEVEMENTS_KEY, None)
        # 项目背景保留完整原文；项目要点保留 Markdown 层级。
        out.append(item)
    # 技能/兴趣分组需要方向（tags）元数据，因此直接用原始选中条目（含 tags）
    if block == "skills":
        return compact_skills(selected)
    if block == "interests":
        return compact_interests(selected)
    # GitHub star 徽章：与 server/lib/github-stars.js 行为一致，读本机缓存注入项目名
    if block == "projects":
        inject_github_stars(out)
    # 非 GitHub 品牌链接：自动选择 lobe-icons 品牌 logo
    if block == "projects":
        inject_url_brands(out)
    return out


GITHUB_STARS_CACHE = os.path.join(
    os.path.expanduser("~"), ".resume-manager", "github-stars.json"
)


def parse_github_repo_url(url):
    if not url or not isinstance(url, str):
        return None
    s = url.strip()
    if s.lower().endswith(".git"):
        s = s[:-4]
    s = s.rstrip("/")
    m = re.search(r"github\.com/([^/?#]+/[^/?#]+)", s, re.IGNORECASE)
    if not m:
        return None
    parts = m.group(1).split("/")
    if len(parts) != 2:
        return None
    if not all(re.fullmatch(r"[\w.-]+", p) for p in parts):
        return None
    return parts[0] + "/" + parts[1]


def format_star_count(n):
    try:
        num = int(n)
    except (TypeError, ValueError):
        return ""
    if num >= 1000000:
        s = "%.1f" % (num / 1000000.0)
        return s.rstrip("0").rstrip(".") + "m"
    if num >= 1000:
        s = "%.1f" % (num / 1000.0)
        return s.rstrip("0").rstrip(".") + "k"
    return str(num)


def load_stars_cache():
    try:
        with open(GITHUB_STARS_CACHE, "r", encoding="utf-8") as f:
            data = json.load(f)
        return data if isinstance(data, dict) else {}
    except Exception:
        return {}


def inject_github_stars(items):
    cache = load_stars_cache()
    for it in items:
        owner_repo = parse_github_repo_url(it.get("url"))
        hit = cache.get(owner_repo) if owner_repo else None
        if not hit:
            continue
        try:
            count = int(hit.get("count"))
        except (TypeError, ValueError):
            continue
        if count < 0:
            continue  # 无效值不显示；0 star 仅显示仓库地址
        badge = format_star_count(count) if count > 0 else ""
        if count > 0 and not badge:
            continue
        if "[github|" in str(it.get("name") or ""):
            continue
        it["name"] = "%s [github|%s|%s]" % (it["name"], owner_repo, badge)


def inject_url_brands(items):
    """非 GitHub 的品牌链接（OpenAI/Claude/DeepSeek 等）追加品牌 logo 标记"""
    for it in items:
        name = str(it.get("name") or "")
        if "[github|" in name or "[brand|" in name:
            continue
        brand = detect_url_brand(it.get("url"))
        if brand:
            it["name"] = "%s [brand|%s]" % (name, brand["iconId"])


def apply_layout_fonts(layout, fonts):
    family = get_typography_font_family(
        layout.get("engine"),
        layout.get("template"),
        normalize_font_settings(fonts),
    )
    if family:
        layout["typography"] = {**(layout.get("typography") or {}), "fontFamily": family}
    return layout


def build_layout(v, defaults):
    d = dict(defaults.get("layout") or {})
    d.update(v.get("layout") or {})
    order = v.get("sectionOrder") or defaults.get("sectionOrder")
    if order:
        d.setdefault("sections", {})["order"] = list(order)
    if d.get("engine") == "latex":
        d["advanced"] = {"showUrls": False, **(d.get("advanced") or {})}
    return apply_layout_fonts(d, v.get("fonts"))


def build_layouts(v, defaults):
    """多引擎支持：v.htmlLayout 存在时同时输出 HTML 布局（PDF + HTML 一次构建）"""
    latex = build_layout(v, defaults)
    html = v.get("htmlLayout")
    if html and html.get("template"):
        html_layout = {"engine": "html", "template": html["template"]}
        if isinstance(html.get("typography"), dict):
            html_layout["typography"] = dict(html["typography"])
        order = v.get("sectionOrder") or defaults.get("sectionOrder")
        if order:
            html_layout.setdefault("sections", {})["order"] = list(order)
        return [latex, apply_layout_fonts(html_layout, v.get("fonts"))]
    return [latex]


def compose_variant(name, v, defaults):
    content = {}
    for block, cfg in (v.get("blocks") or {}).items():
        if block == "basics":
            b = compose_basics()
            if b:
                content["basics"] = b
        else:
            items = compose_list(block, cfg)
            if items:
                content[block] = items
    return {
        "content": content,
        "locale": {"language": v.get("locale", defaults.get("locale", "zh-hans"))},
        "layouts": build_layouts(v, defaults),
    }


def main():
    args = sys.argv[1:]
    check_only = "--check" in args
    only = [a for a in args if not a.startswith("--")]

    cfg = load(VARIANTS_FILE)
    defaults = cfg.get("defaults") or {}
    variants = cfg.get("variants") or {}
    if not variants:
        print("variants.yml 中没有定义任何变体")
        return 1

    os.makedirs(OUT_DIR, exist_ok=True)
    generated = []
    for name, v in variants.items():
        if only and name not in only:
            continue
        resume = compose_variant(name, v, defaults)
        path = os.path.join(OUT_DIR, name + ".yml")
        if not check_only:
            with open(path, "w", encoding="utf-8") as f:
                f.write(HEADER)
                yaml.dump(resume, f, Dumper=Dumper, allow_unicode=True,
                          sort_keys=False, default_flow_style=False, width=120)
        blocks = list(resume["content"].keys())
        print(f"[{'check' if check_only else 'gen '}] {name}.yml  sections={blocks}")
        generated.append(path)
    if check_only:
        print(f"check passed: {len(generated)} variant(s) would be generated")
    else:
        print(f"generated {len(generated)} file(s) in {os.path.relpath(OUT_DIR, ROOT)}/")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: server/lib/compose.js
// 组合引擎：按 scripts/variants.yml 配方从 data/ 信息全集动态组稿
// 与私有数据仓 templates/private-repo/scripts/compose.py 语义一致（JS 版），
// 数据格式规范见 docs/DATA-FORMAT.md
import fs from 'node:fs'
import path from 'node:path'
import yaml from 'js-yaml'
import { readCategory, dataFile, META_KEYS } from './data-store.js'
import { getSettings } from '../config.js'
import { loadStarsCache, parseGithubRepoUrl, formatStarCount } from './github-stars.js'
import { getTypographyFontFamily, normalizeFontSettings } from './font-options.js'

const ACHIEVEMENTS_KEY = 'achievements'
const VARIANTS_FILE = (repo) => path.join(repo, 'scripts', 'variants.yml')
const OUT_DIR = (repo) => path.join(repo, 'resumes')

const HEADER = [
  '# 本文件由 resume-manager 组合引擎生成，请勿手改。',
  '# 改信息 -> data/；改筛选/模板/章节顺序 -> scripts/variants.yml',
  '---',
  '',
].join('\n')

export function loadVariantsDoc(repo) {
  const file = VARIANTS_FILE(repo)
  if (!fs.existsSync(file)) return { defaults: {}, variants: {} }
  return yaml.load(fs.readFileSync(file, 'utf8')) || { defaults: {}, variants: {} }
}

function toSummaryString(summary) {
  if (summary == null) return null
  if (typeof summary === 'string') return summary
  if (Array.isArray(summary)) {
    const lines = summary.filter(Boolean).map((s) => `- ${s}`)
    return lines.length ? lines.join('\n') : null
  }
  return String(summary)
}

function normalizeProjectSummary(summary) {
  const text = toSummaryString(summary)?.trim()
  if (!text) return null
  // 已使用 Markdown 列表时原样保留缩进；旧纯文本按行迁移为顶级项目要点。
  if (/^\s{0,3}(?:[-+*]|\d+[.)])\s+/m.test(text)) return text
  return text
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => `- ${line}`)
    .join('\n')
}

function normalizeProjectBackground(description) {
  const text = String(description || '').trim()
  if (!text) return null
  return /^项目背景[：:]/.test(text) ? text : `项目背景：${text}`
}

function tagOverlap(itemTags, wanted) {
  if (!wanted || wanted.length === 0) return true
  if (!itemTags || itemTags.length === 0) return false
  return wanted.some((t) => itemTags.includes(t))
}

function selectEntries(entries, cfg) {
  if (!entries || entries.length === 0) return []
  if (!cfg || cfg.include === 'all') return [...entries]
  const ids = cfg.ids
  const tags = cfg.tags
  const result = []
  for (const e of entries) {
    if (Array.isArray(ids)) {
      if (ids.includes(e.id)) result.push(e)
      continue
    }
    if (tagOverlap(e.tags, tags)) result.push(e)
  }
  return result
}

function collectAchievements(entry, wanted) {
  if (!(ACHIEVEMENTS_KEY in entry)) return null
  const items = []
  for (const a of entry[ACHIEVEMENTS_KEY] || []) {
    if (a && typeof a === 'object') {
      // 无标签成就 = 通用，保留；有标签的必须与方向命中
      if (wanted && wanted.length && a.tags && a.tags.length && !tagOverlap(a.tags, wanted)) continue
      if (a.text) items.push(String(a.text))
    } else if (a != null) {
      items.push(String(a))
    }
  }
  return items.length ? items.map((t) => `- ${t}`).join('\n') : null
}

function buildSummary(entry, wanted) {
  if (!(ACHIEVEMENTS_KEY in entry)) return entry.summary
  return collectAchievements(entry, wanted)
}

function compactSkills(items) {
  // 按细分方向（tags）分组：每个方向一行“方向：技能、技能”，跨方向技能同时出现在多行。
  const dirOrder = []
  const groups = new Map()
  for (const item of items) {
    const candidates = Array.isArray(item.keywords) && item.keywords.length ? item.keywords : [item.name]
    const names = candidates
      .map((c) => String(c || '').replace(/^(熟悉|掌握)\s*/, '').trim())
      .filter(Boolean)
    if (!names.length) continue
    const tags = Array.isArray(item.tags)
      ? item.tags.filter((t) => typeof t === 'string' && t.trim())
      : []
    const dirs = tags.length ? tags : ['通用']
    for (const dir of dirs) {
      if (!groups.has(dir)) {
        groups.set(dir, [])
        dirOrder.push(dir)
      }
      const list = groups.get(dir)
      for (const n of names) if (!list.includes(n)) list.push(n)
    }
  }
  return dirOrder
    .filter((dir) => (groups.get(dir) || []).length)
    .map((dir) => ({
      name: `${dir}：${groups.get(dir).join('、')}`,
      // schema 要求 level 为合法枚举；构建阶段会移除模板追加的“专家/大师”。
      level: 'Expert',
      keywords: [],
    }))
}

function compactInterests(items) {
  const names = []
  for (const item of items) {
    const name = String(item.name || '').trim()
    if (name && !names.includes(name)) names.push(name)
  }
  return names.length ? [{ name: names.join('、'), keywords: [] }] : []
}

function composeBasics(repo) {
  const b = { ...readCategory(repo, 'basics') }
  // photo 是管理端构建元数据；YAMLResume basics schema 不接受该字段，由渲染后处理注入。
  delete b.photo
  b.summary = toSummaryString(b.summary)
  return Object.keys(b).length ? b : null
}

function composeList(repo, block, cfg) {
  const entries = readCategory(repo, block)
  const selected = selectEntries(entries, cfg)
  if (!selected.length) return null
  const wanted = (cfg && cfg.tags) || []
  const out = []
  for (const e of selected) {
    const item = {}
    for (const [k, v] of Object.entries(e)) {
      // subtitle/source/role/stage 为管理端字段（副标题/来源/职责/阶段），schema 无此字段，不进简历
      if (k !== 'subtitle' && k !== 'source' && k !== 'role' && k !== 'stage' && !META_KEYS.has(k) && !k.startsWith('_') && k !== ACHIEVEMENTS_KEY) item[k] = v
    }
    // 字段别名：work 章节 schema 字段名是 name（数据里用 company 更自然）
    if ('company' in item && !('name' in item)) {
      item.name = item.company
      delete item.company
    }
    if (block === 'projects') {
      // 字段映射：数据 background → schema description；数据 tech → schema keywords
      if ('background' in item) {
        item.description = item.background
        delete item.background
      }
      if ('tech' in item) {
        item.keywords = item.tech
        delete item.tech
      }
      const summary = normalizeProjectSummary(e.summary)
      const achievements = collectAchievements(e, wanted)
      if (summary || achievements) {
        item.summary = [summary, achievements ? `**成果**\n${achievements}` : null].filter(Boolean).join('\n\n')
      } else {
        delete item.summary
      }
      const background = normalizeProjectBackground(item.description)
      if (background) item.description = background
      else delete item.description
    } else {
      const summary = buildSummary(e, wanted)
      if (summary != null) item.summary = toSummaryString(summary)
      else if (ACHIEVEMENTS_KEY in e) delete item.summary
    }
    // 项目背景保留完整原文；项目要点保留 Markdown 层级。
    out.push(item)
  }
  // 技能/兴趣分组需要方向（tags）元数据，因此直接用原始选中条目（含 tags）
  if (block === 'skills') return compactSkills(selected)
  if (block === 'interests') return compactInterests(selected)
  // GitHub star 徽章：正式发布前已刷新缓存，组合时统一读缓存（预览不发起网络请求）
  if (block === 'projects' && getSettings().starsEnabled !== false) injectGithubStars(out)
  return out
}

// 从本地缓存读取 star 数，将「[github|owner/repo|4.2k]」追加到项目名后
// （仅 GitHub 仓库链接且缓存命中；渲染器转为 GitHub Logo + 地址 + stars 徽章）
function injectGithubStars(items) {
  const cache = loadStarsCache()
  for (const it of items) {
    const ownerRepo = parseGithubRepoUrl(it.url)
    const hit = ownerRepo && cache[ownerRepo]
    const count = Number(hit?.count)
    if (!Number.isFinite(count) || count < 0) continue // 无缓存或无效值不显示；0 star 仅显示仓库地址
    const badge = count > 0 ? formatStarCount(count) : ''
    if (count > 0 && !badge) continue
    if (String(it.name || '').includes('[github|')) continue
    it.name = `${it.name} [github|${ownerRepo}|${badge}]`
  }
}

function applyLayoutFonts(layout, fonts) {
  const family = getTypographyFontFamily(layout.engine, layout.template, normalizeFontSettings(fonts))
  if (!family) return layout
  layout.typography = { ...(layout.typography || {}), fontFamily: family }
  return layout
}

function buildLayout(v, defaults) {
  const d = { ...((defaults && defaults.layout) || {}) }
  Object.assign(d, v.layout || {})
  const order = v.sectionOrder || (defaults && defaults.sectionOrder)
  if (Array.isArray(order) && order.length) {
    d.sections = { ...(d.sections || {}), order: [...order] }
  }
  if (d.engine === 'latex') d.advanced = { showUrls: false, ...(d.advanced || {}) }
  return applyLayoutFonts(d, v.fonts)
}

// 多引擎支持：v.htmlLayout（如 {engine:'html', template:'calm'}）存在时，
// 同时输出 HTML 布局 → yamlresume build 会同时产出 PDF 与 HTML
function buildLayouts(v, defaults) {
  const latex = buildLayout(v, defaults)
  const html = v.htmlLayout
  if (html && html.template) {
    const htmlLayout = { engine: 'html', template: html.template }
    if (html.typography && typeof html.typography === 'object') htmlLayout.typography = { ...html.typography }
    // html 引擎章节顺序同样生效
    const order = v.sectionOrder || (defaults && defaults.sectionOrder)
    if (Array.isArray(order) && order.length) {
      htmlLayout.sections = { ...(htmlLayout.sections || {}), order: [...order] }
    }
    return [latex, applyLayoutFonts(htmlLayout, v.fonts)]
  }
  return [latex]
}

export function composeVariant(repo, name, v, defaults) {
  const content = {}
  for (const [block, cfg] of Object.entries(v.blocks || {})) {
    if (block === 'basics') {
      const b = composeBasics(repo)
      if (b) content.basics = b
    } else {
      const items = composeList(repo, block, cfg)
      if (items) content[block] = items
    }
  }
  return {
    content,
    locale: { language: v.locale || (defaults && defaults.locale) || 'zh-hans' },
    layouts: buildLayouts(v, defaults),
  }
}

export function listVariants(repo) {
  const doc = loadVariantsDoc(repo)
  const defaults = doc.defaults || {}
  const names = Object.keys(doc.variants || {})
  return names.map((name) => {
    const v = doc.variants[name]
    // 统计每个方向将命中多少条目（预览用）
    const counts = {}
    for (const [block, cfg] of Object.entries(v.blocks || {})) {
      if (block === 'basics') continue
      const n = selectEntries(readCategory(repo, block), cfg).length
      if (n > 0) counts[block] = n
    }
    const fonts = normalizeFontSettings(v.fonts)
    return { name, ...v, fonts: Object.keys(fonts).length ? fonts : undefined, matched: counts }
  })
}

export function saveVariantsDoc(repo, doc) {
  const file = VARIANTS_FILE(repo)
  const clean = structuredClone(doc || { defaults: {}, variants: {} })
  clean.defaults = clean.defaults || {}
  clean.variants = clean.variants || {}
  for (const variant of Object.values(clean.variants)) {
    if (!variant || typeof variant !== 'object') continue
    delete variant.name
    delete variant.label
    delete variant.branch
    delete variant.matched
  }
  fs.mkdirSync(path.dirname(file), { recursive: true })
  fs.writeFileSync(file, yaml.dump(clean, { noRefs: true, lineWidth: -1, sortKeys: false }), 'utf8')
}

export function generateVariant(repo, name, variant, defaults = {}) {
  if (!variant) return null
  fs.mkdirSync(OUT_DIR(repo), { recursive: true })
  const resume = composeVariant(repo, name, variant, defaults)
  const text = HEADER + yaml.dump(resume, { noRefs: true, lineWidth: -1, sortKeys: false })
    .replace(/: null\s*$/gm, ':')
  fs.writeFileSync(path.join(OUT_DIR(repo), `${name}.yml`), text, 'utf8')
  return name
}

export function generateAll(repo, only) {
  const doc = loadVariantsDoc(repo)
  const defaults = doc.defaults || {}
  const variants = doc.variants || {}
  const names = only ? [only] : Object.keys(variants)
  const generated = []
  for (const name of names) {
    if (generateVariant(repo, name, variants[name], defaults)) generated.push(name)
  }
  return generated
}

export { dataFile }
